runtime-modules/project-token/src/types.rs

```diff
@@ -76,42 +137,325 @@ pub enum TransferPolicy<Hash> {
+    /// Account that acts as the source of the tokens on sale
+    pub tokens_source: AccountId,
+    /// Token's unit price in JOY
+    pub unit_price: JoyBalance,
+    /// Number of tokens on sale
+    pub upper_bound_quantity: Balance,
+    /// Optional block in the future when the sale should start (by default: starts immediately)
+    pub starts_at: Option<BlockNumber>,
+    /// Sale duration in blocks
+    pub duration: BlockNumber,
+    /// Optional vesting schedule for all tokens on sale
+    pub vesting_schedule: Option<VestingScheduleParams>,
+    /// Optional total sale purchase amount cap per member
+    pub cap_per_member: Option<Balance>,
+    /// Optional sale metadata
+    pub metadata: Option<Vec<u8>>,
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, Default)]
+pub struct TokenSale<JoyBalance, Balance, BlockNumber, VestingScheduleParams, AccountId> {
+    /// Token's unit price in JOY
+    pub unit_price: JoyBalance,
+    /// Number of tokens still on sale (if any)
+    pub quantity_left: Balance,
+    /// Account that acts as the source of the tokens on sale
+    pub tokens_source: AccountId,
+    /// Block at which the sale started / will start
+    pub start_block: BlockNumber,
+    /// Sale duration (in blocks)
+    pub duration: BlockNumber,
+    /// Optional vesting schedule for all tokens on sale
+    pub vesting_schedule: Option<VestingScheduleParams>,
+    /// Optional total sale purchase amount cap per member
+    pub cap_per_member: Option<Balance>,
+}
+
+impl<JoyBalance, Balance, BlockNumber, AccountId>
+    TokenSale<JoyBalance, Balance, BlockNumber, VestingScheduleParams<BlockNumber>, AccountId>
+where
+    BlockNumber: Saturating + Zero + Copy + Clone + PartialOrd,
+    Balance: Saturating + Clone + Copy + From<u32> + Unsigned + TryInto<u32> + TryInto<u64> + Ord,
+{
+    pub(crate) fn try_from_params<T: Trait>(
+        params: TokenSaleParamsOf<T>,
+    ) -> Result<TokenSaleOf<T>, DispatchError> {
+        let current_block = <frame_system::Module<T>>::block_number();
+        let start_block = params.starts_at.unwrap_or(current_block);
+
+        ensure!(
+            start_block >= current_block,
+            Error::<T>::SaleStartingBlockInThePast
+        );
+
+        Ok(TokenSale {
+            start_block,
+            duration: params.duration,
+            unit_price: params.unit_price,
+            quantity_left: params.upper_bound_quantity,
+            vesting_schedule: params.vesting_schedule,
+            tokens_source: params.tokens_source,
+            cap_per_member: params.cap_per_member,
+        })
+    }
+
+    pub(crate) fn get_vesting_schedule(
+        &self,
+        amount: Balance,
+    ) -> VestingSchedule<BlockNumber, Balance> {
+        self.vesting_schedule.as_ref().map_or(
+            // Default VestingSchedule when none specified (distribute all tokens right after sale ends)
+            VestingSchedule::<BlockNumber, Balance> {
```

Comment:
Why is this default impotent schedule needed? it should not be.

Reply:
It was supposed to serve as a way to delay the distribution of tokens until the sale is finished, accordingly to this comment:
> - all sale terms will be uniform across all partcipants: quantity cap, price, cliff, unlocking rate
> - **cashout only happens in the last block of the sale.**
> - **vesting and cliff only applies from that point on**

_Source: https://github.com/Joystream/joystream/issues/2362#issuecomment-1075221896_

Now, I didn't know that the unvested tokens would be eligible for staking at the time I implemented this, so now I can see a flaw in this approach which is that the tokens bought on sale can be immediately staked, which gives an advantage to those who bought earlier into the sale, however I'm not sure what the current expectations are w.r.t. that and whether something changed since the comment I quoted was published.

Reply:
> It was supposed to serve as a way to delay the distribution of tokens until the sale is finished

Ok, not totally clear how it does this, but may be a moot point now, see next remark.

> cashout only happens in the last block of the sale.

Oh, I had forgotten that, is there any benefit to this actually? This make UX much worse, in that people must first buy then cashout. What does this buy us? if nothing, then certainly lets drop.

> I didn't know that the unvested tokens would be eligible for staking at the time I implemented this, so now I can see a flaw in this approach an advantage to those who bought earlier into the sale

Anyone arbing a $CRT, whether via OTC, AMM, an external DEX/CEX, has to be aware of what the underlying $JOY pool attached to it, and dynamics around an upcoming/ongoing revenue split. Seems this is part of what people need to be aware of when transacting, and that for thisr eason that revenue splits must be slow to both start and end.

Do you think this is a big issue?

Reply:
> Oh, I had forgotten that, is there any benefit to this actually?

I can't think of anything specific, besides the point that it kind of "evens the playing field" for all sale participants, but even that is not entirely the case, since early participants still have the benefit of using the bought tokens immediately for staking, as noted before.

The benefit on the codebase side is that since all sale purchases are subject to **some** vesting, it allows us to sidestep the need for introducing another storage map like `sale_purchase_by_token_by_member` or some other, separate way to track the amount of tokens already purchased on the current sale by a given member (since we can just use `acccount.vesting_schedules` for this, thanks to indexing by `VestingSource`).

So thinking about this, in order to allow instant cashouts we'd need one of the following:
- `sale_purchase_by_member_by_token` which will be cleaned with `remove_prefix` upon sale completion. I'm guessing `remove_prefix` does not have a deterministic cost (?), so perhaps there's also a need to either introduce max. number participants per sale and use worst-case-scenario in benchmarking or force the caller of sale finalization extrinsic to provide the map size as extrinsic argument and calculate the cost base on that
- `last_sale_total_purchased_amount` as part of `account_data`, which will be a `(sale_id, amount)` pair and it just won't be cleaned up, only updated.

Reply:
# Questions

1. Is it the case that we need one of these two fixes in order to enforce the per member cap that applies? 
2. Is there any other reason?
3. When reading the code I forget now: why do we actually need to track the origin of a vesting schedule using `VestingSource` ? Determining which, if any, are no longer in effect, does not depend on this, but on how much they are locking at the current block. Is it so that we can keep updating the same vesting schedule from a single sale multiple times if the buyer purchases more, without having to add a new schedule?

# Remarks

- `remove_prefix` is deterministic, its just a function of `N`, which is controlled by the users, and thus may end up becoming impossible to ever perform within the weight constraint of a block.
- `last_sale_total_purchased_amount` is by far the best solution.

# Conclusion

- I'm finally seeing now why the first schedule was being made now. :D
- While the sale UX is worse, I also see that it is a bit more orderly to give everyone access to their purchased tokens at the same time. However, lets not forget that one can simulate delayed cashouts by manually specifying such a schedule, even if chain allows instant cashouts! so this solution is in a way the most general
- The instant cashouts require new storage variable `last_sale_total_purchased_amount`, while delayed cashouts is worse UX and also is a bit weird in introducing reusing the concept of vesting for two very distinct sort of activties: 
- **In my mind, its a really close call, I think you should decide what is best here, but I lean towards instant cashouts**.